addons/mail/controllers/attachment.py

```diff
@@ -65,50 +64,28 @@ def mail_attachment_upload(self, ufile, thread_id, thread_model, is_pending=Fals
                     "res_model": "mail.compose.message",
                 }
             )
-        if request.env.user.share:
-            # Only generate the access token if absolutely necessary (= not for internal user).
-            vals["access_token"] = request.env["ir.attachment"]._generate_access_token()
         try:
             # sudo: ir.attachment - posting a new attachment on an accessible thread
             attachment = request.env["ir.attachment"].sudo().create(vals)
             attachment._post_add_create(**kwargs)
-            res = {"data": Store(attachment, extra_fields="access_token").get_result()}
+            res = {"data": Store(attachment, extra_fields="as_author_access_token").get_result()}
         except AccessError:
             res = {"error": _("You are not allowed to upload an attachment here.")}
         return request.make_json_response(res)
 
     @http.route("/mail/attachment/delete", methods=["POST"], type="jsonrpc", auth="public")
     @add_guest_to_context
-    def mail_attachment_delete(self, attachment_id, access_token=None, **kwargs):
+    def mail_attachment_delete(self, attachment_id, access_token=None):
         attachment = request.env["ir.attachment"].browse(int(attachment_id)).exists()
         if not attachment:
             request.env.user._bus_send("ir.attachment/delete", {"id": attachment_id})
             return
-        attachment_message = request.env["mail.message"].sudo().search(
-            [("attachment_ids", "in", attachment.ids)], limit=1)
-        message = ThreadController._get_message_with_access(attachment_message.id, mode="create", **kwargs)
-        if not request.env.user.share:
-            # Check through standard access rights/rules for internal users.
-            attachment._delete_and_notify(message)
-            return
-        # For non-internal users 2 cases are supported:
-        #   - Either the attachment is linked to a message: verify the request is made by the author of the message (portal user or guest).
-        #   - Either a valid access token is given: also verify the message is pending (because unfortunately in portal a token is also provided to guest for viewing others' attachments).
-        # sudo: ir.attachment: access is validated below with membership of message or access token
-        attachment_sudo = attachment.sudo()
-        if message:
-            if not ThreadController._can_delete_attachment(message, **kwargs):
-                raise NotFound()
-        else:
-            if (
-                not access_token
-                or not attachment_sudo.access_token
-                or not consteq(access_token, attachment_sudo.access_token)
-            ):
-                raise NotFound()
-            if attachment_sudo.res_model != "mail.compose.message" or attachment_sudo.res_id != 0:
-                raise NotFound()
-        attachment_sudo._delete_and_notify(message)
+        attachment._check_attachments_access([access_token])
```

Comment:
Note: not for this PR, but I don't like the API of this method.
We have to not forget to raise when it is problematic, rather than simply returning True when it is acceptable, which makes it more error prone.
Also it's not trivial when reading code to know the route is protected because the method will raise.